internal/pkg/agent/application/dispatcher/dispatcher.go

```diff
@@ -192,70 +203,165 @@ func detectTypes(actions []fleetapi.Action) []string {
+		for _, action := range ad.queue.Actions() {
+			if _, ok := action.(*fleetapi.ActionUpgrade); !ok {
+				continue
+			}
+
+			if !yield(action.ID(), action) {
+				return
+			}
+		}
+	})
+
 	for i := len(actions) - 1; i >= 0; i-- {
 		action := actions[i]
-		// If it is a cancel action, remove from list and dispatch
-		if action.Type() == fleetapi.ActionTypeCancel {
-			actions = append(actions[:i], actions[i+1:]...)
-			if err := ad.dispatchAction(ctx, action, acker); err != nil {
-				ad.log.Errorf("Unable to dispatch cancel action id %s: %v", action.ID(), err)
-			}
+
+		cancelAction, ok := action.(*fleetapi.ActionCancel)
+		if !ok {
+			continue
+		}
+
+		actions = append(actions[:i], actions[i+1:]...)
+		if err := ad.dispatchAction(ctx, action, acker); err != nil {
+			ad.log.Errorf("Unable to dispatch cancel action id %s: %v", action.ID(), err)
+		}
+
+		if _, exists := queuedUpgradeActions[cancelAction.Data.TargetID]; exists {
+			*upgradeDetailsNeedUpdate = true
 		}
 	}
 	return actions
 }
 
-// gatherQueuedActions will dequeue actions from the action queue and separate those that have already expired.
-func (ad *ActionDispatcher) gatherQueuedActions(ts time.Time) (queued, expired []fleetapi.Action) {
-	actions := ad.queue.DequeueActions()
-	for _, action := range actions {
+// mergeWithQueuedActions will dequeue actions from the action queue, excluding those that have already expired, and merge
+// them with the passed actions. If upgrade actions are found to inside the dispatchable action, upgradeDetailsNeedUpdate
+// will be set to false, otherwise if upgrade actions are found to be expired, upgradeDetailsNeedUpdate will be set to true.
+func (ad *ActionDispatcher) mergeWithQueuedActions(ts time.Time, actions []fleetapi.Action, upgradeDetailsNeedUpdate *bool) []fleetapi.Action {
+	var expired []fleetapi.ScheduledAction
+	var expiredUpgradeActions []fleetapi.ScheduledAction
+	dequeuedActions := ad.queue.DequeueActions()
+
+	for _, action := range dequeuedActions {
 		exp, _ := action.Expiration()
 		if ts.After(exp) {
-			expired = append(expired, action)
+			if action.Type() == fleetapi.ActionTypeUpgrade {
+				// this is an expired upgrade action thus we need to recalculate the upgrade details
+				*upgradeDetailsNeedUpdate = true
+				expiredUpgradeActions = append(expiredUpgradeActions, action)
+			} else {
+				expired = append(expired, action)
+			}
+
+			continue
+		}
+		actions = append(actions, action)
+	}
+
+	ad.log.Debugf("Gathered %d actions from queue, %d non-upgrade actions expired, %d upgrade actions expired",
+		len(dequeuedActions), len(expired), len(expiredUpgradeActions))
+	ad.log.Debugf("Expired non-upgrade actions: %v", expired)
+	ad.log.Debugf("Expired upgrade actions (won't be removed from the queue): %v", expiredUpgradeActions)
+
+	for _, expiredUpgradeAction := range expiredUpgradeActions {
+		startTime, err := expiredUpgradeAction.StartTime()
```

Comment:
Can you add a comment here saying that at this point in the code there really should be a `StartTime` on the action, because these are expired upgrade actions. I just want to make warnings in the code clear, because before this change it was not clear.

Reply:
have a look here 6f75b29b772af12fe38091555902a056c14b9f35